js/models/messages.js

```diff
@@ -1900,33 +1912,63 @@
       return message;
     },
 
-    async handleDataMessage(initialMessage, confirm) {
-      // This function is called from the background script in a few scenarios:
-      //   1. on an incoming message
-      //   2. on a sent message sync'd from another device
-      //   3. in rare cases, an incoming message can be retried, though it will
-      //      still go through one of the previous two codepaths
-      const ourNumber = textsecure.storage.user.getNumber();
-      const message = this;
-      const source = message.get('source');
-      const type = message.get('type');
-      let conversationId = message.get('conversationId');
-      const authorisation = await libloki.storage.getGrantAuthorisationForSecondaryPubKey(
-        source
+    async handleSecondaryDeviceFriendRequest(pubKey) {
+      // fetch the device mapping from the server
+      const deviceMapping = await lokiFileServerAPI.getUserDeviceMapping(
+        pubKey
       );
-      const primarySource =
-        (authorisation && authorisation.primaryDevicePubKey) || source;
-      const isGroupMessage = !!initialMessage.group;
-      if (isGroupMessage) {
-        conversationId = initialMessage.group.id;
-      } else if (source !== ourNumber && authorisation) {
-        // Ignore auth from our devices
-        conversationId = authorisation.primaryDevicePubKey;
+      if (!deviceMapping) {
+        return false;
+      }
+      // Only handle secondary pubkeys
+      if (deviceMapping.isPrimary === '1' || !deviceMapping.authorisations) {
+        return false;
+      }
+      const { authorisations } = deviceMapping;
+      // Secondary devices should only have 1 authorisation from a primary device
+      if (authorisations.length !== 1) {
+        return false;
+      }
+      const authorisation = authorisations[0];
+      if (!authorisation) {
+        return false;
+      }
+      if (!authorisation.grantSignature) {
+        return false;
+      }
+      const isValid = await libloki.crypto.validateAuthorisation(authorisation);
+      if (!isValid) {
+        return false;
+      }
+      const correctSender = pubKey === authorisation.secondaryDevicePubKey;
+      if (!correctSender) {
+        return false;
       }
+      const { primaryDevicePubKey } = authorisation;
+      // ensure the primary device is a friend
+      const c = window.ConversationController.get(primaryDevicePubKey);
+      if (!c || !c.isFriendWithAnyDevice()) {
+        return false;
+      }
+      await libloki.storage.savePairingAuthorisation(authorisation);
 
-      const GROUP_TYPES = textsecure.protobuf.GroupContext.Type;
+      return true;
+    },
 
+    /**
+     * Returns true if the message is already completely handled and confirmed
+     * and the processing of this message must stop.
+     */
+    handleGroupMessage(source, initialMessage, primarySource, confirm) {
```

Comment:
each of those new methods (split from huge `handleDataMessage`) can return true.
if that's the case, it means the processing of that single message is completed and `handleDataMessage` must return early

Reply:
yes, exactly if the test before the next function is valid, we had the same logic before.
we returned null only on some specific cases otherwise the message was following its path: see here for example: https://github.com/loki-project/session-desktop/blob/clearnet/js/models/messages.js#L1995

Reply:
👍